File: packages/less/src/less/tree/variable.test.js
import { describe, it, expect, beforeEach } from 'vitest';
import Variable from './variable';
import Call from './call';
import Node from './node';

describe('Variable', () => {
    let mockContext;
    let mockFileInfo;

    beforeEach(() => {
        mockFileInfo = {
            filename: 'test.less'
        };
        mockContext = {
            frames: [
                {
                    functionRegistry: {
                        get: () => null
                    }
                }
            ],
            importantScope: [{ important: false }],
            inCalc: false,
            enterCalc: () => {},
            exitCalc: () => {},
            mathOn: true
        };
    });

    describe('Variable creation', () => {
        it('should create a Variable instance with correct properties', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            expect(variable.name).toBe('@color');
            expect(variable._index).toBe(1);
            expect(variable._fileInfo).toBe(mockFileInfo);
            expect(variable.type).toBe('Variable');
        });

        it('should handle undefined parameters', () => {
            const variable = new Variable('@color');
            expect(variable.name).toBe('@color');
            expect(variable._index).toBeUndefined();
            expect(variable._fileInfo).toBeUndefined();
            expect(variable.type).toBe('Variable');
        });
    });

    describe('Variable evaluation', () => {
        it('should evaluate a simple variable', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            const mockValue = { eval: () => ({ value: 'red' }) };
            mockContext.frames = [
                {
                    functionRegistry: {
                        get: () => null
                    },
                    variable: (name) =>
                        name === '@color' ? { value: mockValue } : null
                }
            ];

            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'red' });
        });

        it('should search through multiple frames', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            const mockValue = { eval: () => ({ value: 'blue' }) };
            mockContext.frames = [
                {
                    functionRegistry: { get: () => null },
                    variable: () => null
                },
                {
                    functionRegistry: { get: () => null },
                    variable: (name) =>
                        name === '@color' ? { value: mockValue } : null
                }
            ];

            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'blue' });
        });

        it('should handle variables with @@ prefix', () => {
            const variable = new Variable('@@var', 1, mockFileInfo);
            const innerValue = { eval: () => ({ value: 'color' }) };
            const outerValue = { eval: () => ({ value: 'red' }) };

            mockContext.frames = [
                {
                    functionRegistry: {
                        get: () => null
                    },
                    variable: (name) => {
                        if (name === '@var') return { value: innerValue };
                        if (name === '@color') return { value: outerValue };
                        return null;
                    }
                }
            ];

            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'red' });
        });

        it('should handle multiple @@ prefix levels', () => {
            const variable = new Variable('@@@var', 1, mockFileInfo);
            const firstValue = { eval: () => ({ value: 'color' }) };
            const secondValue = { eval: () => ({ value: 'primary' }) };
            const finalValue = { eval: () => ({ value: 'blue' }) };

            mockContext.frames = [
                {
                    functionRegistry: {
                        get: () => null
                    },
                    variable: (name) => {
                        if (name === '@var') return { value: firstValue };
                        if (name === '@color') return { value: secondValue };
                        if (name === '@primary') return { value: finalValue };
                        return null;
                    }
                }
            ];

            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'blue' });
        });

        it('should propagate file info to evaluated result', () => {
            const sourceFileInfo = { filename: 'source.less' };
            const variable = new Variable('@color', 42, sourceFileInfo);

            // Create a Node instance that will properly receive file info
            class MockNode extends Node {
                constructor() {
                    super();
                    this.value = 'red';
                    this._fileInfo = sourceFileInfo;
                    this._index = 42;
                }
            }
            const resultNode = new MockNode();

            const mockValue = {
                eval: () => resultNode
            };
            mockContext.frames = [
                {
                    functionRegistry: { get: () => null },
                    variable: () => ({ value: mockValue })
                }
            ];

            const result = variable.eval(mockContext);
            expect(result.fileInfo()).toEqual(sourceFileInfo);
            expect(result.getIndex()).toBe(42);
        });

        it('should throw error for recursive variable definition', () => {
            const variable = new Variable('@recursive', 1, mockFileInfo);
            variable.evaluating = true;

            expect(() => variable.eval(mockContext)).toThrowError(
                'Recursive variable definition for @recursive'
            );
        });

        it('should throw error for undefined variable', () => {
            const variable = new Variable('@undefined', 1, mockFileInfo);
            mockContext.frames = [
                {
                    functionRegistry: {
                        get: () => null
                    },
                    variable: () => null
                }
            ];

            expect(() => variable.eval(mockContext)).toThrowError(
                'variable @undefined is undefined'
            );
        });

        it('should handle variables in calc context', () => {
            const variable = new Variable('@number', 1, mockFileInfo);
            const mockValue = {
                eval: () => ({ value: 42 })
            };
            mockContext.inCalc = true;
            mockContext.frames = [
                {
                    functionRegistry: {
                        get: () => null
                    },
                    variable: (name) =>
                        name === '@number' ? { value: mockValue } : null
                }
            ];

            const result = variable.eval(mockContext);
            expect(result).toBeInstanceOf(Call);
            expect(result.name).toBe('_SELF');
            expect(result.args[0]).toEqual({ value: 42 });
        });

        it('should handle variables with important flag', () => {
            const variable = new Variable('@important', 1, mockFileInfo);
            const mockValue = { eval: () => ({ value: 'important!' }) };
            mockContext.frames = [
                {
                    functionRegistry: {
                        get: () => null
                    },
                    variable: (name) =>
                        name === '@important'
                            ? { value: mockValue, important: true }
                            : null
                }
            ];

            variable.eval(mockContext);
            expect(mockContext.importantScope[0].important).toBe(true);
        });
    });

    describe('find method', () => {
        it('should find first matching item in array', () => {
            const variable = new Variable('@test', 1, mockFileInfo);
            const arr = [1, 2, 3, 4, 5];
            const result = variable.find(arr, (item) =>
                item === 3 ? item : null
            );
            expect(result).toBe(3);
        });

        it('should return null if no match found', () => {
            const variable = new Variable('@test', 1, mockFileInfo);
            const arr = [1, 2, 3, 4, 5];
            const result = variable.find(arr, (item) =>
                item === 6 ? item : null
            );
            expect(result).toBeNull();
        });

        it('should handle empty array', () => {
            const variable = new Variable('@test', 1, mockFileInfo);
            const result = variable.find([], (item) => item);
            expect(result).toBeNull();
        });

        it('should handle non-array input gracefully', () => {
            const variable = new Variable('@test', 1, mockFileInfo);
            const result = variable.find('not an array', () => false);
            expect(result).toBeNull();
        });
    });

    describe('Edge cases for variable names', () => {
        it('should handle empty variable name', () => {
            const variable = new Variable('', 1, mockFileInfo);
            mockContext.frames = [
                {
                    variable: () => null
                }
            ];
            // Empty variable names should throw undefined variable error
            expect(() => variable.eval(mockContext)).toThrowError(
                'variable  is undefined'
            );
        });

        it('should handle special characters in variable names', () => {
            const variable = new Variable('@special!#$%', 1, mockFileInfo);
            const mockValue = { eval: () => ({ value: 'special' }) };
            mockContext.frames = [
                {
                    variable: (name) =>
                        name === '@special!#$%' ? { value: mockValue } : null
                }
            ];
            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'special' });
        });

        it('should handle very long variable names', () => {
            const longName = '@' + 'a'.repeat(1000);
            const variable = new Variable(longName, 1, mockFileInfo);
            const mockValue = { eval: () => ({ value: 'long' }) };
            mockContext.frames = [
                {
                    variable: (name) =>
                        name === longName ? { value: mockValue } : null
                }
            ];
            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'long' });
        });
    });

    describe('Context variations', () => {
        it('should handle empty frames array', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            mockContext.frames = [];
            expect(() => variable.eval(mockContext)).toThrowError(
                'variable @color is undefined'
            );
        });

        it('should respect frame precedence with duplicate variables', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            mockContext.frames = [
                {
                    variable: () => ({
                        value: { eval: () => ({ value: 'red' }) }
                    })
                },
                {
                    variable: () => ({
                        value: { eval: () => ({ value: 'blue' }) }
                    })
                }
            ];
            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'red' });
        });

        it('should handle importantScope with important flag', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            // Ensure importantScope has at least one element
            mockContext.importantScope = [{ important: false }];
            const mockValue = { eval: () => ({ value: 'red' }) };
            mockContext.frames = [
                {
                    variable: () => ({ value: mockValue, important: true })
                }
            ];
            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'red' });
            // Verify the important flag was set
            expect(mockContext.importantScope[0].important).toBe(true);
        });

        it('should handle importantScope without important flag', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            mockContext.importantScope = [{ important: false }];
            const mockValue = { eval: () => ({ value: 'red' }) };
            mockContext.frames = [
                {
                    variable: () => ({ value: mockValue })
                }
            ];
            const result = variable.eval(mockContext);
            expect(result).toEqual({ value: 'red' });
            // Verify the important flag remains unchanged
            expect(mockContext.importantScope[0].important).toBe(false);
        });
    });

    describe('Error handling', () => {
        it('should handle invalid fileInfo', () => {
            const variable = new Variable('@color', 1, null);
            mockContext.frames = [
                {
                    variable: () => ({
                        value: { eval: () => ({ value: 'red' }) }
                    })
                }
            ];
            expect(() => variable.eval(mockContext)).not.toThrow();
        });

        it('should handle null context', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            expect(() => variable.eval(null)).toThrowError();
        });

        it('should handle malformed variable values', () => {
            const variable = new Variable('@color', 1, mockFileInfo);
            mockContext.frames = [
                {
                    variable: () => ({ value: null })
                }
            ];
            expect(() => variable.eval(mockContext)).toThrowError();
        });
    });
});

File: packages/less/src/less/go_parser/variable_test.go
package go_parser

import (
	"strings"
	"testing"
)

type mockContext struct {
	frames         []Frame
	importantScope []map[string]bool
	inCalc         bool
	mathOn         bool
}

func (m *mockContext) GetFrames() []Frame {
	return m.frames
}

func (m *mockContext) GetImportantScope() []map[string]bool {
	return m.importantScope
}

func (m *mockContext) IsInCalc() bool {
	return m.inCalc
}

func (m *mockContext) IsMathOn() bool {
	return m.mathOn
}

func (m *mockContext) SetMathOn(on bool) {
	m.mathOn = on
}

func (m *mockContext) EnterCalc() {
	m.inCalc = true
}

func (m *mockContext) ExitCalc() {
	m.inCalc = false
}

type mockFrame struct {
	vars map[string]map[string]any
}

func (m *mockFrame) Variable(name string) map[string]any {
	return m.vars[name]
}

type mockValue struct {
	value any
}

func (m *mockValue) Eval(context EvalContext) (any, error) {
	return map[string]any{"value": m.value}, nil
}

func TestVariable(t *testing.T) {
	mockFileInfo := map[string]any{
		"filename": "test.less",
	}

	t.Run("Variable creation", func(t *testing.T) {
		t.Run("should create a Variable instance with correct properties", func(t *testing.T) {
			variable := NewVariable("@color", 1, mockFileInfo)
			if variable.name != "@color" {
				t.Errorf("Expected name to be @color, got %s", variable.name)
			}
			if variable._index != 1 {
				t.Errorf("Expected index to be 1, got %d", variable._index)
			}
			if variable._fileInfo["filename"] != mockFileInfo["filename"] {
				t.Error("Expected fileInfo filename to match mockFileInfo filename")
			}
			if variable.GetType() != "Variable" {
				t.Errorf("Expected type to be Variable, got %s", variable.GetType())
			}
		})

		t.Run("should handle undefined parameters", func(t *testing.T) {
			variable := NewVariable("@color", 0, nil)
			if variable.name != "@color" {
				t.Errorf("Expected name to be @color, got %s", variable.name)
			}
			if variable._index != 0 {
				t.Errorf("Expected index to be 0, got %d", variable._index)
			}
			if variable._fileInfo != nil {
				t.Error("Expected fileInfo to be nil")
			}
		})
	})

	t.Run("Variable evaluation", func(t *testing.T) {
		t.Run("should evaluate a simple variable", func(t *testing.T) {
			variable := NewVariable("@color", 1, mockFileInfo)
			mockValue := &mockValue{value: "red"}
			context := &mockContext{
				frames: []Frame{
					&mockFrame{
						vars: map[string]map[string]any{
							"@color": {"value": mockValue},
						},
					},
				},
				importantScope: []map[string]bool{{"important": false}},
			}

			result, err := variable.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result.(map[string]any)["value"] != "red" {
				t.Errorf("Expected value to be red, got %v", result.(map[string]any)["value"])
			}
		})

		t.Run("should handle variables with @@ prefix", func(t *testing.T) {
			variable := NewVariable("@@var", 1, mockFileInfo)
			innerValue := &mockValue{value: "color"}
			outerValue := &mockValue{value: "red"}
			context := &mockContext{
				frames: []Frame{
					&mockFrame{
						vars: map[string]map[string]any{
							"@var":   {"value": innerValue},
							"@color": {"value": outerValue},
						},
					},
				},
				importantScope: []map[string]bool{{"important": false}},
			}

			result, err := variable.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result.(map[string]any)["value"] != "red" {
				t.Errorf("Expected value to be red, got %v", result.(map[string]any)["value"])
			}
		})

		t.Run("should throw error for recursive variable definition", func(t *testing.T) {
			variable := NewVariable("@recursive", 1, mockFileInfo)
			variable.evaluating = true

			context := &mockContext{
				frames: []Frame{&mockFrame{vars: map[string]map[string]any{}}},
				importantScope: []map[string]bool{{"important": false}},
			}

			_, err := variable.Eval(context)
			if err == nil {
				t.Error("Expected error for recursive variable definition")
			}
		})

		t.Run("should throw error for undefined variable", func(t *testing.T) {
			variable := NewVariable("@undefined", 1, mockFileInfo)
			context := &mockContext{
				frames: []Frame{&mockFrame{vars: map[string]map[string]any{}}},
				importantScope: []map[string]bool{{"important": false}},
			}

			_, err := variable.Eval(context)
			if err == nil {
				t.Error("Expected error for undefined variable")
			}
		})

		t.Run("should handle variables in calc context", func(t *testing.T) {
			variable := NewVariable("@number", 1, mockFileInfo)
			mockValue := &mockValue{value: 42}
			context := &mockContext{
				frames: []Frame{
					&mockFrame{
						vars: map[string]map[string]any{
							"@number": {"value": mockValue},
						},
					},
				},
				importantScope: []map[string]bool{{"important": false}},
				inCalc:         true,
			}

			result, err := variable.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}

			// In calc context, result should be a Call with _SELF
			call, ok := result.(*Call)
			if !ok {
				t.Fatalf("Expected result to be *Call, got %T", result)
			}
			if call.Name != "_SELF" {
				t.Errorf("Expected call name to be _SELF, got %s", call.Name)
			}
			if len(call.Args) != 1 {
				t.Errorf("Expected call to have 1 argument, got %d", len(call.Args))
			}
			// The argument should be our mockValue
			if call.Args[0] != mockValue {
				t.Errorf("Expected call argument to be mockValue")
			}
		})

		t.Run("should handle variables with important flag", func(t *testing.T) {
			variable := NewVariable("@important", 1, mockFileInfo)
			mockValue := &mockValue{value: "important!"}
			context := &mockContext{
				frames: []Frame{
					&mockFrame{
						vars: map[string]map[string]any{
							"@important": {
								"value":     mockValue,
								"important": true,
							},
						},
					},
				},
				importantScope: []map[string]bool{{"important": false}},
			}

			_, err := variable.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if !context.importantScope[0]["important"] {
				t.Error("Expected important flag to be set to true")
			}
		})

		t.Run("should handle multiple @@ prefix levels", func(t *testing.T) {
			variable := NewVariable("@@@var", 1, mockFileInfo)
			firstValue := &mockValue{value: "color"}
			secondValue := &mockValue{value: "primary"}
			finalValue := &mockValue{value: "blue"}
			context := &mockContext{
				frames: []Frame{
					&mockFrame{
						vars: map[string]map[string]any{
							"@var":     {"value": firstValue},
							"@color":   {"value": secondValue},
							"@primary": {"value": finalValue},
						},
					},
				},
				importantScope: []map[string]bool{{"important": false}},
			}

			result, err := variable.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result.(map[string]any)["value"] != "blue" {
				t.Errorf("Expected value to be blue, got %v", result.(map[string]any)["value"])
			}
		})

		t.Run("Edge cases for variable names", func(t *testing.T) {
			t.Run("should handle empty variable name", func(t *testing.T) {
				variable := NewVariable("", 1, mockFileInfo)
				context := &mockContext{
					frames: []Frame{&mockFrame{vars: map[string]map[string]any{}}},
					importantScope: []map[string]bool{{"important": false}},
				}

				_, err := variable.Eval(context)
				if err == nil {
					t.Error("Expected error for empty variable name")
				}
			})

			t.Run("should handle special characters in variable names", func(t *testing.T) {
				variable := NewVariable("@special!#$%", 1, mockFileInfo)
				mockValue := &mockValue{value: "special"}
				context := &mockContext{
					frames: []Frame{
						&mockFrame{
							vars: map[string]map[string]any{
								"@special!#$%": {"value": mockValue},
							},
						},
					},
					importantScope: []map[string]bool{{"important": false}},
				}

				result, err := variable.Eval(context)
				if err != nil {
					t.Errorf("Unexpected error: %v", err)
				}
				if result.(map[string]any)["value"] != "special" {
					t.Errorf("Expected value to be special, got %v", result.(map[string]any)["value"])
				}
			})

			t.Run("should handle very long variable names", func(t *testing.T) {
				longName := "@" + strings.Repeat("a", 1000)
				variable := NewVariable(longName, 1, mockFileInfo)
				mockValue := &mockValue{value: "long"}
				context := &mockContext{
					frames: []Frame{
						&mockFrame{
							vars: map[string]map[string]any{
								longName: {"value": mockValue},
							},
						},
					},
					importantScope: []map[string]bool{{"important": false}},
				}

				result, err := variable.Eval(context)
				if err != nil {
					t.Errorf("Unexpected error: %v", err)
				}
				if result.(map[string]any)["value"] != "long" {
					t.Errorf("Expected value to be long, got %v", result.(map[string]any)["value"])
				}
			})
		})
	})
} 